yt_dlp/extractor/crunchyroll.py

```diff
@@ -70,28 +73,35 @@ def _update_auth(self, lang):
             'locale': ('localization', 'locale'),
         }) or None
 
-        if self._AUTH_HEADERS:
+        if self._BASIC_AUTH:
             return
 
-        if self.is_logged_in:
-            grant_type, key = 'etp_rt_cookie', 'accountAuthClientId'
-        else:
-            grant_type, key = 'client_id', 'anonClientId'
-
         app_config = self._search_json(r'__APP_CONFIG__\s*=', webpage, 'app config', None)
+        cx_api_param = app_config['cxApiParams']['accountAuthClientId' if self.is_logged_in else 'anonClientId']
+        self.write_debug(f'Using cxApiParam={cx_api_param}')
+        self._BASIC_AUTH = 'Basic ' + base64.b64encode(f'{cx_api_param}:'.encode()).decode()
+
+    def _update_auth(self):
+        if self._AUTH_HEADERS and self._AUTH_REFRESH > time_seconds():
+            return
+
+        assert self._BASIC_AUTH, '_update_query needs to be called at least one time beforehand'
+        grant_type = 'etp_rt_cookie' if self.is_logged_in else 'client_id'
         auth_response = self._download_json(
             f'{self._BASE_URL}/auth/v1/token', None, note=f'Authenticating with grant_type={grant_type}',
-            headers={
-                'Authorization': 'Basic ' + base64.b64encode(('%s:' % app_config['cxApiParams'][key]).encode()).decode()
-            }, data=f'grant_type={grant_type}'.encode())
+            headers={'Authorization': self._BASIC_AUTH}, data=f'grant_type={grant_type}'.encode())
 
         self._AUTH_HEADERS = {'Authorization': auth_response['token_type'] + ' ' + auth_response['access_token']}
+        self._AUTH_REFRESH = time_seconds(seconds=traverse_obj(auth_response, ('expires_in', {float_or_none}), default=300))
```

Comment:
Maybe leave some buffer time here for slop in communications? I believe the webapp renews 10 seconds before the nominal expiry time.